pkg/approver/manager/subjectaccessreview.go

```diff
@@ -0,0 +1,191 @@
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package evaluator
+
+import (
+	"context"
+	"fmt"
+	"sort"
+	"strings"
+
+	cmapi "github.com/jetstack/cert-manager/pkg/apis/certmanager/v1"
+	authzv1 "k8s.io/api/authorization/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	cmpapi "github.com/cert-manager/policy-approver/pkg/apis/policy/v1alpha1"
+	"github.com/cert-manager/policy-approver/pkg/approver"
+)
+
+var _ Interface = &subjectaccessreview{}
+
+// subjectaccessreview is a manager that calls evaluators with
+// CertificateRequestPolicys that have been RBAC bound to the user who appears
+// in the passed CertificateRequest to Evaluate.
+type subjectaccessreview struct {
+	client client.Client
+
+	evaluators []approver.Evaluator
+}
+
+// NewSubjectAccessReview constructs a new approver Manager which evaluates
+// whether CertificateRequests should be approved or denied, managing
+// registered evaluators.
+func NewSubjectAccessReview(client client.Client, evaluators []approver.Evaluator) *subjectaccessreview {
+	return &subjectaccessreview{
+		client:     client,
+		evaluators: evaluators,
+	}
+}
+
+// Review will evaluate whether the incoming CertificateRequest should be
+// approved. All evaluators will be called with CertificateRequestPolicys that
+// have been RBAC bound to the user included in the CertificateRequest.
+func (s *subjectaccessreview) Review(ctx context.Context, cr *cmapi.CertificateRequest) (ReviewResponse, error) {
+	crps := new(cmpapi.CertificateRequestPolicyList)
+	if err := s.client.List(ctx, crps); err != nil {
+		return ReviewResponse{}, err
+	}
+
+	// If no CertificateRequestPolicies exist in the cluster, return
+	// ResultUnprocessed.
+	if len(crps.Items) == 0 {
+		return ReviewResponse{Result: ResultUnprocessed, Message: "No CertificateRequestPolicies exist"}, nil
+	}
+
+	boundPolicies, err := s.boundPolicies(ctx, cr, crps.Items)
+	if err != nil {
+		return ReviewResponse{}, fmt.Errorf("failed to determine bound policies: %w", err)
+	}
+
+	// If no policies are bound to the requesting user, return denied.
+	if len(boundPolicies) == 0 {
+		return ReviewResponse{
+			Result:  ResultDenied,
+			Message: "No CertificateRequestPolicies bound or applicable",
+		}, nil
+	}
+
+	// policyMessages hold the aggregated messages of each evaluator response,
+	// keyed by the policy name that was executed.
+	var policyMessages []policyMessage
+
+	// Run every evaluators against ever policy which is bound to the requesting
+	// user.
+	for _, crp := range boundPolicies {
+		var (
+			evaluatorDenied   bool
+			evaluatorMessages []string
+		)
+
+		for _, evaluator := range s.evaluators {
```

Comment:
Suggestion: log which evaluator we're evaluating against (just so that users can 'see' that any custom evaluator they've registered is actually used